packages/asset-swapper/src/utils/market_operation_utils/types.ts

```diff
@@ -273,17 +278,65 @@ export interface GetMarketOrdersOpts {
  * A composable operation the be run in `DexOrderSampler.executeAsync()`.
  */
 export interface BatchedOperation<TResult> {
-    encodeCall(contract: ERC20BridgeSamplerContract): string;
-    handleCallResultsAsync(contract: ERC20BridgeSamplerContract, callResults: string): Promise<TResult>;
+    encodeCall(): string;
+    handleCallResultsAsync(callResults: string): Promise<TResult>;
 }
 
 export interface SourceQuoteOperation<TFillData extends FillData = FillData>
-    extends BatchedOperation<Array<Quote<TFillData>>> {
-    source: ERC20BridgeSource;
+    extends BatchedOperation<BigNumber[]>,
+        SourceInfo<TFillData> {
+    readonly source: ERC20BridgeSource;
+}
+
+type Parameters<T> = T extends (...args: infer TArgs) => any ? TArgs : never;
+
+export interface SamplerContractCall<
+    TFunc extends (...args: any[]) => ContractFunctionObj<any>,
+    TFillData extends FillData = FillData
+> {
+    contract: ERC20BridgeSamplerContract;
+    function: TFunc;
+    params: Parameters<TFunc>;
+    callback?: (callResults: string, fillData: TFillData) => Promise<BigNumber[]>;
 }
 
-export interface OptimizedOrdersAndQuoteReport {
+export class SamplerContractOperation<
+    TFunc extends (...args: any[]) => ContractFunctionObj<any>,
+    TFillData extends FillData = FillData
+> implements SourceQuoteOperation<TFillData> {
+    public readonly source: ERC20BridgeSource;
+    public fillData: TFillData;
+    private readonly _samplerContract: ERC20BridgeSamplerContract;
+    private readonly _samplerFunction: TFunc;
+    private readonly _params: Parameters<TFunc>;
+    private readonly _callback?: (callResults: string, fillData: TFillData) => Promise<BigNumber[]>;
+
+    constructor(opts: SourceInfo<TFillData> & SamplerContractCall<TFunc, TFillData>) {
+        this.source = opts.source;
+        this.fillData = opts.fillData || ({} as TFillData); // tslint:disable-line:no-object-literal-type-assertion
+        this._samplerContract = opts.contract;
+        this._samplerFunction = opts.function;
+        this._params = opts.params;
+        this._callback = opts.callback;
+    }
+
+    public encodeCall(): string {
+        return this._samplerFunction
+            .bind(this._samplerContract)(...this._params)
+            .getABIEncodedTransactionData();
+    }
+    public async handleCallResultsAsync(callResults: string): Promise<BigNumber[]> {
```

Comment:
do we have anything that is async anymore? Offchain Bancor/Balancer are out of the usual sampler operations?

It does scare me if this async fn can execute for a long time :P

Reply:
yep, you're right